# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/HarisAhmed786/FlyRank-Scalable-SEO-Ranking-ML-Agent/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [19]:
import os
import sys
import subprocess
from pathlib import Path

REPO_DIR = "flyrank-ml-internship-starter"
REPO_URL = "https://github.com/HarisAhmed786/FlyRank-Scalable-SEO-Ranking-ML-Agent"

# Always return to /content first in Colab
os.chdir("/content")

# Clone only if it does not exist
if not os.path.isdir(REPO_DIR):
    subprocess.run(
        ["git", "clone", "--depth", "1", REPO_URL, REPO_DIR],
        check=True
    )

# Enter the repo exactly once
os.chdir(REPO_DIR)

print("Working directory:", os.getcwd())
print(
    "Dataset exists:",
    Path("data/raw/content_refresh_anonymized.csv").exists()
)

Working directory: /content/flyrank-ml-internship-starter
Dataset exists: True


## 1. My lane (or freestyle) and why
I am choosing the content-refresh prioritization lane: identify pages that show signals associated with search impression decline so SEO editors can prioritize which pages to review first. This lane is useful because the dataset contains observed decline outcomes and allows us to investigate which signals are associated with pages needing attention.


## 2. The question: decision, action, cost of a wrong call

This work improves the decision of which page to refresh or audit next. SEO/content editors act on the output by reviewing the top-ranked pages first and choosing which content to update. A wrong recommendation wastes editor time on low-risk pages and may allow truly declining pages to keep losing traffic, so the cost is both wasted effort and missed recovery opportunity.

In [20]:
import pandas as pd
from pathlib import Path

root = Path.cwd()
print(f"Starting search from: {root}")
while not (root / 'data' / 'raw' / 'content_refresh_anonymized.csv').exists() and root != root.parent:
    print(f"Checking: {root / 'data' / 'raw' / 'content_refresh_anonymized.csv'}")
    root = root.parent

data_path = root / 'data' / 'raw' / 'content_refresh_anonymized.csv'
print(f"Final data_path used: {data_path}")
df = pd.read_csv(data_path)

declining_rate = (df['trend_direction'] == 'down').mean()
print(f'Rows: {len(df)}')
print(f'Declining pages: {declining_rate:.3%} of the sample')
print('Declining pages are the observed target for the refresh decision.')

Starting search from: /content/flyrank-ml-internship-starter
Final data_path used: /content/flyrank-ml-internship-starter/data/raw/content_refresh_anonymized.csv
Rows: 30000
Declining pages: 54.207% of the sample
Declining pages are the observed target for the refresh decision.


## 3. Quick look at the data (2-3 real numbers)

Over half of the sample pages (about 54.2%) are labeled as declining, so this is a common operational issue. The dataset has an average of about 5,200 search impressions over 90 days, meaning declines affect pages with real traffic. A simple signal like keyword search volume has almost no correlation with actual impressions, so ranking pages for refresh review requires more than one rule.

Another observation from exploration was that declining pages had a lower median content age (216 days) compared with stable and growing pages (around 292–300 days), suggesting that age alone is not a reliable refresh signal.

In [21]:
import pandas as pd
from pathlib import Path

root = Path.cwd()
print(f"Starting search from: {root}")
while not (root / 'data' / 'raw' / 'content_refresh_anonymized.csv').exists() and root != root.parent:
    print(f"Checking: {root / 'data' / 'raw' / 'content_refresh_anonymized.csv'}")
    root = root.parent

data_path = root / 'data' / 'raw' / 'content_refresh_anonymized.csv'
print(f"Final data_path used: {data_path}")
df = pd.read_csv(data_path)

declining_rate = (df['trend_direction'] == 'down').mean()
print(f'Declining share: {declining_rate:.3%}')
print('Mean 90-day impressions:', round(df['impressions_90d'].mean()))
print('Search volume vs impressions correlation: {:.3f}'.format(df['search_volume'].corr(df['impressions_90d'])))

Starting search from: /content/flyrank-ml-internship-starter
Final data_path used: /content/flyrank-ml-internship-starter/data/raw/content_refresh_anonymized.csv
Declining share: 54.207%
Mean 90-day impressions: 5200
Search volume vs impressions correlation: 0.001


## 4. Careful words: what I can and can't claim

This work can say what is observed in the anonymized data: which pages are currently in a declining impression trend, and which features support ranking them for refresh review. It can support editorial decision-making, and it can show directional patterns. It cannot claim causal proof about Google, it cannot say it is predicting Google’s algorithm, and it cannot claim that a page will definitely recover after a refresh.

In [22]:
import pandas as pd
from pathlib import Path

root = Path.cwd()
print(f"Starting search from: {root}")
while not (root / 'data' / 'raw' / 'content_refresh_anonymized.csv').exists() and root != root.parent:
    print(f"Checking: {root / 'data' / 'raw' / 'content_refresh_anonymized.csv'}")
    root = root.parent

data_path = root / 'data' / 'raw' / 'content_refresh_anonymized.csv'
print(f"Final data_path used: {data_path}")
df = pd.read_csv(data_path)

down_rate = (df['trend_direction'] == 'down').mean()
print(f'Observed decline label share: {down_rate:.3%}')
print('The target is observed from the data, and this work supports decision-making rather than causal claims.')
print('Do not use trend_direction or trend_pct as model features because they are label sources.')

Starting search from: /content/flyrank-ml-internship-starter
Final data_path used: /content/flyrank-ml-internship-starter/data/raw/content_refresh_anonymized.csv
Observed decline label share: 54.207%
The target is observed from the data, and this work supports decision-making rather than causal claims.
Do not use trend_direction or trend_pct as model features because they are label sources.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.